### DISCUSSION TITLE
Sanitizing Quotes in PR Titles

### DISCUSSION THREAD

**Discussion by colintheshots-meetup** [community/community#27065] [Actions]
<p>I want to script our release automation to handle PR titles and descriptions. However, when a PR’s title or description contains any single or double quotes or other control characters like backticks, the shell interprets those quotes and causes the action to fail.</p>
<p>Here’s an example of the problem I’m talking about. If the PR title or description contains a quote, this action fails. By the time sed gets access to sanitize the input, it’s already too late and the Github Action has terminated.</p>
<p>Is there any mechanism I can use to sanitize the PR body and title input before the step reaches a syntax error?</p>
<pre><code class="lang-auto">name: On PR Merged into Develop

on:
  pull_request:
    types: [closed]
    branches:
      - main

jobs:
  test:
    if: github.event.pull_request.merged
    runs-on: ubuntu-latest
    steps:
      - name: strip quotes from PR body and title
        run: |
          echo 'PR_BODY&lt;&lt;EOF' &gt;&gt; $GITHUB_ENV
          echo '${{github.event.pull_request.body}}' | sed -e 's/[\"\']//g' &gt;&gt; $GITHUB_ENV
          echo 'EOF' &gt;&gt; $GITHUB_ENV
          echo 'PR_TITLE&lt;&lt;EOF' &gt;&gt; $GITHUB_ENV
          echo '${{github.event.pull_request.title}}' | sed -e 's/[\"\']//g' &gt;&gt; $GITHUB_ENV
          echo 'EOF' &gt;&gt; $GITHUB_ENV
      - name: show variables
        run: cat $GITHUB_ENV
</code></pre>
<p>Before anyone asks, the built-in <code>env</code> setters suffer from the exact same problem, which was why I was trying to sanitize input in the first place.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>I don’t think there’s a built-in way to do that. What I’d do is read the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context">file <code>github.event_path</code> points at</a> with a JSON parser and sanitize the fields from there, without first having to pass the strings through the shell (or any other programming language).</p>
<p>For example, to get a shell-escaped version of the pull request title using Python in the <code>PR_TITLE</code> environment variable:</p>
<pre><code class="lang-yaml">      - name: escape pull request title
        run: |
          import json
          import os
          import shlex
          with open('${{ github.event_path }}') as fh:
              event = json.load(fh)
          escaped = shlex.quote(event['pull_request']['title'])
          with open(os.environ['GITHUB_ENV'], 'a') as fh:
              print(f'PR_TITLE={escaped}', file=fh)
        shell: python3 {0}
</code></pre>
<p>Maybe the documentation should actually warn about directly using user-supplied data in workflows… :thinking:</p>

**Comment by jamesckemp** [comment]
<p>Hey, this has got me very close to a solution, but I’m seeing an issue when I echo the contents of <code>$PR_TITLE</code> to a file.</p>
<p><code>echo $PR_TITLE &gt;&gt; file.md</code></p>
<p>Results in:</p>
<p><code>'Test '"'"'single'"'"' "double"'</code></p>
<p>Should I be using something other than <code>echo</code> to append the variable contents to a file?</p>

**Comment by airtower-luna** [comment]
<p>Could you share a bit more about how your workflow and the inputs look?</p>
<p>In most cases <code>echo</code> and a redirect is good to write to files, but you call it through the shell, which means any shell escapes are processed at that point. I used Python above to ensure the escapes actually make it into the file.</p>

**Comment by jamesckemp** [comment]
<p>Sure thing.</p>
<p>When a pull request is merged, I’m adding the pull request title to the end of a markdown file (changelog). I want to cover bases where the PR title has single or double quotes, but also backticks.</p>
<p>These are the two steps doing the processing:</p>
<pre><code class="lang-auto">- name: Create Changelog Entry String
  run: |
    import json
    import os
    import shlex
    with open('${{ github.event_path }}') as fh:
        event = json.load(fh)
    escaped = shlex.quote(event['pull_request']['title'])
    with open(os.environ['GITHUB_ENV'], 'a') as fh:
        print(f'PR_TITLE={escaped}', file=fh)
  shell: python3 {0}

# Update file.
- run: |
    echo $PR_TITLE &gt;&gt; changelog.md
</code></pre>
<p>And this is converting this: <code>Test 'single' "double"</code></p>
<p>To this in the file: <code>'Test '"'"'single'"'"' "double"'</code></p>

**Comment by jamesckemp** [comment]
<p>I think removing the <code>shlex.quote()</code> method solves the issue for me!</p>
<pre><code class="lang-auto">      - name: Create Changelog Entry String
        run: |
          import json
          import os
          with open('${{ github.event_path }}') as fh:
              event = json.load(fh)
          escaped = event['pull_request']['title']
          with open(os.environ['GITHUB_ENV'], 'a') as fh:
              print(f'PR_TITLE={escaped}', file=fh)
        shell: python3 {0}
      # Update changelog.
      - run: |
          printf "%b" "$PR_TITLE  \n" &gt;&gt; changelog.md
</code></pre>
<p>This is producing the correct output. Does anything look dangerous?</p>

**Comment by airtower-luna** [comment]
<p>It depends on what kind of escaping you need, so if you don’t need a shell-escaped version of the title that should work. But if you just want to append to the changelog, why not do it directly from the Python fragment, without the detour via <code>GITHUB_ENV</code>? That’d eliminate any change of interference.</p>

**Comment by glasser** [comment]
According to the docs, loading it into an env var first works:

https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable

**Reply by rbals0445** [reply]
@glasser , perfect answer. THX!!!!!!!!!

**Reply by fortenforge** [reply]
this should be the accepted answer instead of the thing above

**Reply by Noyabronok** [reply]
So easy to escape double quotes this way.  AI kept suggesting other solutions that didn't work.
